scipy/linalg/basic.py

```diff
@@ -814,40 +836,121 @@ def lstsq(a, b, cond=None, overwrite_a=False, overwrite_b=False,
         nrhs = 1
     if m != b1.shape[0]:
         raise ValueError('incompatible dimensions')
-    gelss, = get_lapack_funcs(('gelss',), (a1, b1))
-    if n > m:
+
+    if lapack_driver not in ('gelsd', 'gelsy', 'gelss'):
+        raise ValueError('LAPACK driver "%s"is not found' % lapack_driver)
+
+    lapack_func, lapack_func_lwork_query = get_lapack_funcs((lapack_driver,
+                                        '%s_lwork' % lapack_driver), (a1, b1))
+    # the dtype isn't necessarily the dtype needed for lapack.  It's the 
+    # result of np.find_common_type on the array inputs. 
+    if lapack_func.dtype.char not in 'fdFD':
+        lapack_func.dtype = np.dtype('d')
+    real_data = True if (lapack_func.dtype.kind == 'f') else False
+
+    if m < n:
         # need to extend b matrix as it will be filled with
         # a larger solution matrix
         if len(b1.shape) == 2:
-            b2 = np.zeros((n, nrhs), dtype=gelss.dtype)
+            b2 = np.zeros((n, nrhs), dtype=lapack_func.dtype)
             b2[:m, :] = b1
         else:
-            b2 = np.zeros(n, dtype=gelss.dtype)
+            b2 = np.zeros(n, dtype=lapack_func.dtype)
             b2[:m] = b1
         b1 = b2
 
     overwrite_a = overwrite_a or _datacopied(a1, a)
     overwrite_b = overwrite_b or _datacopied(b1, b)
 
-    # get optimal work array
-    work = gelss(a1, b1, lwork=-1)[4]
-    lwork = work[0].real.astype(int)
-    v, x, s, rank, work, info = gelss(
-        a1, b1, cond=cond, lwork=lwork, overwrite_a=overwrite_a,
-        overwrite_b=overwrite_b)
-
-    if info > 0:
-        raise LinAlgError("SVD did not converge in Linear Least Squares")
-    if info < 0:
-        raise ValueError('illegal value in %d-th argument of internal gelss' %
-                         -info)
-    resids = np.asarray([], dtype=x.dtype)
-    if n < m:
-        x1 = x[:n]
-        if rank == n:
-            resids = np.sum(np.abs(x[n:])**2, axis=0)
-        x = x1
-    return x, resids, rank, s
+    if lapack_driver in ('gelss', 'gelsd'):
+        if cond is None:
+            cond = np.finfo(lapack_func.dtype).eps * 100
+
+        if lapack_driver == 'gelss':
+            # Request of sizes
+            work,info = lapack_func_lwork_query(m,n,nrhs,cond)
+            lwork = np.int(np.real(work))
+            if info != 0:
+                raise ValueError("internal gelss driver lwork query "
+                                 "error, info is %i " % info)
```

Comment:
The above can now be written as `lwork = scipy.linalg.lapack._compute_lwork(lapack_func_lwork_query, m, n, nrhs, cond)`